cpp/src/io/parquet/api.cpp

```diff
@@ -0,0 +1,779 @@
+        column_index_in_read_set(_column_index_in_read_set),
+        //	  row_group_reader(std::move(_row_group_reader)),
+        column_reader(std::move(_column_reader)), column(std::move(_column)),
+        offset(_offset) {}
+};
+
+void
+_ProcessParquetReaderJobsThread(std::vector<ParquetReaderJob> &jobs,
+                                std::mutex &                   lock,
+                                int &                          job_index,
+                                gdf_error &                    gdf_error_out) {
+
+    lock.lock();
+    int current_job = job_index;
+    job_index++;
+    lock.unlock();
+
+    gdf_error current_gdf_error = GDF_SUCCESS;
+
+    while (current_job < jobs.size()) {
+
+        switch (jobs[current_job].column_reader->type()) {
+#define WHEN(TYPE)                                                             \
+    case ::parquet::Type::TYPE: {                                              \
+        std::shared_ptr<gdf::parquet::ColumnReader<                            \
+          ::parquet::DataType<::parquet::Type::TYPE>>>                         \
+          reader = std::static_pointer_cast<gdf::parquet::ColumnReader<        \
+            ::parquet::DataType<::parquet::Type::TYPE>>>(                      \
+            jobs[current_job].column_reader);                                  \
+        if (reader->HasNext()) {                                               \
+            reader->ToGdfColumn(jobs[current_job].column,                      \
+                                jobs[current_job].offset,                      \
+                                jobs[current_job].first_valid_byte,            \
+                                jobs[current_job].last_valid_byte);            \
+        }                                                                      \
+    } break
+            WHEN(BOOLEAN);
+            WHEN(INT32);
+            WHEN(INT64);
+            WHEN(FLOAT);
+            WHEN(DOUBLE);
+        default:
+#ifdef GDF_DEBUG
+            std::cerr << "Column type error from file" << std::endl;
+#endif
+            current_gdf_error =
+              GDF_FILE_ERROR;  //TODO: improve using exception handling
+#undef WHEN
+        }
+
+        lock.lock();
+        if (gdf_error_out != GDF_SUCCESS) {  // if error we want to exit
+            current_job = jobs.size();
+        } else if (current_gdf_error
+                   != GDF_SUCCESS) {  // if error we want to exit
+            gdf_error_out = current_gdf_error;
+            current_job   = jobs.size();
+        } else {
+            current_job = job_index;
+        }
+        job_index++;
+        lock.unlock();
+    }
+}
+
+gdf_error
+_ProcessParquetReaderJobs(std::vector<ParquetReaderJob> &jobs) {
+
+    std::mutex lock;
+    int        job_index     = 0;
+    gdf_error  gdf_error_out = GDF_SUCCESS;
+
+    int num_threads = std::thread::hardware_concurrency();
```

Comment:
With CUDA, this CPU side parallelization should become irrelevant. We should be able to get the same perf with a single thread launching CUDA kernels. Not sure if this is the case ATM.